Add a decomposition for take()

asmeurer opened the pull request:
Presumably this can close https://github.com/pytorch/pytorch/pull/109784

Also related to https://github.com/pytorch/pytorch/issues/93757 (though `take` is not listed there). 

There's no bounds checking here (out of bounds indices cause a segfault or undefined behavior). Should that be added somehow?

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

asmeurer committed:
Add a decomposition for take()

asmeurer committed:
Autoformatter

lezcano reviewed (commented):
Mind posting the triton code it generates? Do so running it with `TORCH_LOGS=output_code`

lezcano commented on torch/_decomp/decompositions.py:
```diff
@@ -4344,6 +4344,17 @@ def _weight_norm_interface(x, y, dim):
+@register_decomposition(aten.take)
+@out_wrapper()
+def take(self, index):
+    flattened = self.reshape(-1)
```
we also have `self.flatten()` which is a bit more explicit, but they both do the same, so whatever.

asmeurer commented:
> Mind posting the triton code it generates? Do so running it with TORCH_LOGS=output_code

```
>>> @torch.compile
... def test(x, idx):
...     return torch.take(x, idx)
>>> test(x, idx)
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG] Output code:
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG] from ctypes import c_void_p, c_long
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG] import torch
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG] import math
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG] import random
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG] import os
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG] import tempfile
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG] from math import inf, nan
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG] from torch._inductor.hooks import run_intermediate_hooks
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG] from torch._inductor.utils import maybe_profile
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG] from torch._inductor.codegen.memory_planning import _align as align
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG] from torch import device, empty, empty_strided
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG] from torch._inductor.codecache import AsyncCompile
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG] from torch._inductor.select_algorithm import extern_kernels
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG] aten = torch.ops.aten
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG] inductor_ops = torch.ops.inductor
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG] assert_size_stride = torch._C._dynamo.guards.assert_size_stride
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG] alloc_from_pool = torch.ops.inductor._alloc_from_pool
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG] reinterpret_tensor = torch.ops.inductor._reinterpret_tensor
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG] async_compile = AsyncCompile()
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG] cpp_fused_take_0 = async_compile.cpp('''
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG] #include "/var/folders/wc/dppcpmxs1tlb36nqcw853wkm0000gn/T/torchinductor_aaronmeurer/2l/c2ljzlm4sosod7u6lyrroqdba6hmfcyijrric6p4t3fhbcmw6osp.h"
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG] extern "C" void kernel(const long* in_ptr0,
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]                        const long* in_ptr1,
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]                        long* out_ptr0)
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG] {
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]     {
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]         #pragma GCC ivdep
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]         for(long x0=static_cast<long>(0L); x0<static_cast<long>(3L); x0+=static_cast<long>(1L))
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]         {
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]             auto tmp0 = in_ptr0[static_cast<long>(x0)];
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]             auto tmp1 = static_cast<long>(0);
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]             auto tmp2 = tmp0 < tmp1;
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]             auto tmp3 = static_cast<long>(10);
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]             auto tmp4 = decltype(tmp0)(tmp0 + tmp3);
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]             auto tmp5 = tmp2 ? tmp4 : tmp0;
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]             auto tmp6 = decltype(tmp5)(tmp5 + 10);
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]             auto tmp7 = tmp5 < 0;
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]             auto tmp8 = tmp7 ? tmp6 : tmp5;
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]             TORCH_CHECK((0 <= tmp8) & (tmp8 < 10L), "index out of bounds: 0 <= tmp8 < 10L")
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]             auto tmp9 = in_ptr1[static_cast<long>(tmp8)];
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]             out_ptr0[static_cast<long>(x0)] = tmp9;
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]         }
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]     }
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG] }
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG] ''')
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG] async_compile.wait(globals())
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG] del async_compile
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG] def call(args):
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]     arg0_1, arg1_1 = args
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]     args.clear()
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]     assert_size_stride(arg0_1, (10, ), (1, ))
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]     assert_size_stride(arg1_1, (3, ), (1, ))
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]     buf0 = empty((3, ), device='cpu', dtype=torch.int64)
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]     cpp_fused_take_0(c_void_p(arg1_1.data_ptr()), c_void_p(arg0_1.data_ptr()), c_void_p(buf0.data_ptr()))
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]     del arg0_1
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]     del arg1_1
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]     return (buf0, )
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG] def benchmark_compiled_module(times=10, repeat=10):
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]     from torch._dynamo.testing import rand_strided
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]     from torch._inductor.utils import print_performance
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]     arg0_1 = rand_strided((10, ), (1, ), device='cpu', dtype=torch.int64)
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]     arg1_1 = rand_strided((3, ), (1, ), device='cpu', dtype=torch.int64)
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]     return print_performance(lambda: call([arg0_1, arg1_1]), times=times, repeat=repeat)
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG] if __name__ == "__main__":
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]     from torch._inductor.wrapper_benchmark import compiled_module_main
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]     compiled_module_main('None', benchmark_compiled_module)
[2023-11-30 08:06:00,486] [1/0] torch._inductor.graph.__output_code: [DEBUG]
[2023-11-30 08:06:00,487] [1/0] torch._inductor.graph.__output_code: [INFO] Output code written to: /var/folders/wc/dppcpmxs1tlb36nqcw853wkm0000gn/T/torchinductor_aaronmeurer/x5/cx5pskplzxrpvbwcee276zpgmcx7vzxcj72if2vznuw73hkijkgt.py
```

If I include `out` I get an error:

```pytb
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/aaronmeurer/Documents/pytorch/torch/_dynamo/eval_frame.py", line 489, in _fn
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/aaronmeurer/Documents/pytorch/torch/_dynamo/eval_frame.py", line 655, in catch_errors
    return callback(frame, cache_entry, hooks, frame_state)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/aaronmeurer/Documents/pytorch/torch/_dynamo/convert_frame.py", line 722, in _convert_frame
    result = inner_convert(frame, cache_entry, hooks, frame_state)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/aaronmeurer/Documents/pytorch/torch/_dynamo/convert_frame.py", line 383, in _convert_frame_assert
    compiled_product = _compile(
                       ^^^^^^^^^
  File "/Users/aaronmeurer/Documents/pytorch/torch/_dynamo/convert_frame.py", line 646, in _compile
    guarded_code = compile_inner(code, one_graph, hooks, transform)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/aaronmeurer/Documents/pytorch/torch/_dynamo/utils.py", line 244, in time_wrapper
    r = func(*args, **kwargs)
        ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/aaronmeurer/Documents/pytorch/torch/_dynamo/convert_frame.py", line 562, in compile_inner
    out_code = transform_code_object(code, transform)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/aaronmeurer/Documents/pytorch/torch/_dynamo/bytecode_transformation.py", line 1033, in transform_code_object
    transformations(instructions, code_options)
  File "/Users/aaronmeurer/Documents/pytorch/torch/_dynamo/convert_frame.py", line 151, in _fn
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/aaronmeurer/Documents/pytorch/torch/_dynamo/convert_frame.py", line 527, in transform
    tracer.run()
  File "/Users/aaronmeurer/Documents/pytorch/torch/_dynamo/symbolic_convert.py", line 2123, in run
    super().run()
  File "/Users/aaronmeurer/Documents/pytorch/torch/_dynamo/symbolic_convert.py", line 818, in run
    and self.step()
        ^^^^^^^^^^^
  File "/Users/aaronmeurer/Documents/pytorch/torch/_dynamo/symbolic_convert.py", line 781, in step
    getattr(self, inst.opname)(inst)
  File "/Users/aaronmeurer/Documents/pytorch/torch/_dynamo/symbolic_convert.py", line 2238, in RETURN_VALUE
    self.output.compile_subgraph(
  File "/Users/aaronmeurer/Documents/pytorch/torch/_dynamo/output_graph.py", line 912, in compile_subgraph
    self.compile_and_call_fx_graph(tx, list(reversed(stack_values)), root)
  File "/Users/aaronmeurer/miniconda3/envs/pytorch/lib/python3.11/contextlib.py", line 81, in inner
    return func(*args, **kwds)
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/aaronmeurer/Documents/pytorch/torch/_dynamo/output_graph.py", line 1080, in compile_and_call_fx_graph
    compiled_fn = self.call_user_compiler(gm)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/aaronmeurer/Documents/pytorch/torch/_dynamo/utils.py", line 244, in time_wrapper
    r = func(*args, **kwargs)
        ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/aaronmeurer/Documents/pytorch/torch/_dynamo/output_graph.py", line 1152, in call_user_compiler
    raise BackendCompilerFailed(self.compiler_fn, e).with_traceback(
  File "/Users/aaronmeurer/Documents/pytorch/torch/_dynamo/output_graph.py", line 1133, in call_user_compiler
    compiled_fn = compiler_fn(gm, self.example_inputs())
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/aaronmeurer/Documents/pytorch/torch/_dynamo/repro/after_dynamo.py", line 117, in debug_wrapper
    compiled_gm = compiler_fn(gm, example_inputs)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/aaronmeurer/Documents/pytorch/torch/__init__.py", line 1657, in __call__
    return compile_fx(model_, inputs_, config_patches=self.config)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/aaronmeurer/Documents/pytorch/torch/_inductor/compile_fx.py", line 1168, in compile_fx
    return aot_autograd(
           ^^^^^^^^^^^^^
  File "/Users/aaronmeurer/Documents/pytorch/torch/_dynamo/backends/common.py", line 55, in compiler_fn
    cg = aot_module_simplified(gm, example_inputs, **kwargs)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/aaronmeurer/Documents/pytorch/torch/_functorch/aot_autograd.py", line 5015, in aot_module_simplified
    compiled_fn = create_aot_dispatcher_function(
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/aaronmeurer/Documents/pytorch/torch/_dynamo/utils.py", line 244, in time_wrapper
    r = func(*args, **kwargs)
        ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/aaronmeurer/Documents/pytorch/torch/_functorch/aot_autograd.py", line 4462, in create_aot_dispatcher_function
    fw_metadata = run_functionalized_fw_and_collect_metadata(
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/aaronmeurer/Documents/pytorch/torch/_functorch/aot_autograd.py", line 1143, in inner
    mutations_under_no_grad_or_inference_mode = mutates_data and are_all_mutations_under_no_grad_or_inference_mode(f_arg)
                                                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/aaronmeurer/Documents/pytorch/torch/_functorch/aot_autograd.py", line 954, in are_all_mutations_under_no_grad_or_inference_mode
    return torch._functionalize_are_all_mutations_under_no_grad_or_inference_mode(t.elem)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/aaronmeurer/Documents/pytorch/torch/__init__.py", line 1927, in __getattr__
    raise AttributeError(f"module '{__name__}' has no attribute '{name}'")
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
AttributeError: module 'torch' has no attribute '_functionalize_are_all_mutations_under_no_grad_or_inference_mode'

Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information


You can suppress this exception and fall back to eager by setting:
    import torch._dynamo
    torch._dynamo.config.suppress_errors = True
```

lezcano commented:
It seems that we are doing the wrapping twice. The issue her is that even if gather does not supports wrapping, `torch.compile`'s implementation does. We should either xfail the tests that test the decomposition with a comment or implement in a different PR wrapping for gather.

asmeurer committed:
Fix HasDecompTest for take

asmeurer committed:
Fix some typos

asmeurer commented:
Here's the result of that benchmark run https://gist.github.com/asmeurer/0dae083265e52973ceb2f7fb5385b407. Note that you have to comment out [this line](https://github.com/pytorch/pytorch/blob/38ae17d166a001ef6837553d1ddffa111624df27/torch/_inductor/lowering.py#L1674) to get the lowering bit to run.

asmeurer commented:
> It seems that we are doing the wrapping twice. The issue her is that even if gather does not supports wrapping, torch.compile's implementation does. We should either xfail the tests that test the decomposition with a comment or implement in a different PR wrapping for gather.

I don't see how to xfail just a single `SampleInput`. Is there an easy way to do it? Of course, I can also just comment the failing test out.

asmeurer commented:
I added benchmarks for both versions to the gist. The duplicate negative index handling doesn't seem to affect the timings. Presumably it just gets compiled away?

lezcano commented:
It doesn't get compiled away. The thing here is that doing the wrapping takes no time compared to loading data from global memory.

Now, the generated code as it stands is not correct, as it would work for things that are more negative than `-size`.

For now, xfail the whole decomp test and leave a comment

asmeurer committed:
Remove negative index handling in the take() decomposition

The eager gather() does not handle negative indices but the compiled gather
does.

lezcano commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -16378,6 +16378,11 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
+           # eager gather() does not support negative indices but compiled
+           # gather does.
+           skips=(
+               DecorateInfo(unittest.expectedFailure),
```
This is skipping all tests. You should just skip the failing tests. See how other OpInfos do it

lezcano commented:
The benchmarks and generated code LGTM. Let's fix the OpInfo and this should be ready to be merged

asmeurer commented:
> Now, the generated code as it stands is not correct, as it would work for things that are more negative than -size.

Ah, good point. Although, as I noted in the OP, this doesn't do any kind of bounds checking and is already undefined/segfaults in that case.

asmeurer committed:
Update XFILs to be more specific

asmeurer commented:
I've updated the opinfos.

asmeurer committed:
Use indexing instead of gather in the take() decomposition

This avoids the issue that gather doesn't not support negative indices.

asmeurer commented on torch/_decomp/decompositions.py:
```diff
@@ -4344,6 +4344,16 @@ def _weight_norm_interface(x, y, dim):
+    # Note: eager gather() does not support negative indices
+    # (https://github.com/pytorch/pytorch/issues/55143), but torch.compile
+    # gather() does.
+    return flattened.gather(dim=0, index=index)
```
Oh nice. I was under the impression from the `gather` suggestion that indexing wouldn't work, but it seems to work just fine.

lezcano commented:
@pytorchbot merge
